src/main/java/edu/agh/wfiis/solid/designpatterns/creational/builder/DesignPattern.java

```diff
@@ -14,66 +11,109 @@
 
     private final String example;
 
-    public DesignPattern(String name, String type, String code, String description, String example) {
-        Objects.requireNonNull(name);
-        Objects.requireNonNull(type);
-        Objects.requireNonNull(code);
+    public static NeedName builder(){
+        return new Builder();
+    }
 
+    private DesignPattern(
+            String name,
+            String type,
+            String code,
+            String description,
+            String example
+    ) {
         this.name = name;
         this.type = type;
         this.code = code;
         this.description = description;
         this.example = example;
     }
 
-    public static Builder builder(){
-        return new Builder();
-    }
-
-    private static class Builder {
-
+    private static class Builder implements BuildingDesignPattern {
         private String name;
-
         private String type;
-
         private String code;
-
         private String description;
-
         private String example;
 
-        public Builder withName(String name) {
+        @Override
+        public NeedType name(String name) {
             this.name = name;
             return this;
         }
 
-        public Builder withType(String type) {
+        @Override
+        public NeedCode type(String type) {
             this.type = type;
             return this;
         }
 
-        public Builder withCode(String code) {
+        @Override
+        public BuildOrExpand code(String code) {
             this.code = code;
             return this;
         }
 
-        public Builder withDescription(String description) {
+        @Override
+        public NeedOptional with() {
+            return this;
+        }
+
+        @Override
+        public BuildOrExpand description(String description) {
             this.description = description;
             return this;
         }
 
-        public Builder withExample(String example) {
+        @Override
+        public BuildOrExpand example(String example) {
             this.example = example;
             return this;
         }
 
+        @Override
         public DesignPattern build() {
-            DesignPattern designPattern = new DesignPattern(name, type, code, description, example);
-            return designPattern;
+            return new DesignPattern(name, type, code, description, example);
         }
     }
 
+    interface BuildingDesignPattern extends NeedName, NeedType, NeedCode, NeedOptional, BuildOrExpand {}
+
+    interface NeedName {
+        NeedType name(String name);
+    }
+
+    interface NeedType {
+        NeedCode type(String type);
+    }
+
+    interface NeedCode {
+        BuildOrExpand code(String code);
+    }
+
+    interface BuildOrExpand {
+        NeedOptional with();
+        DesignPattern build();
+    }
+
+    interface NeedOptional {
```

Comment:
Imo redundant interface, putting optional params directly into BuildOrExpand will be simpler and less obfuscated.